### PR TITLE
Add script to generate metadata for use in compatibility testing. 

### PR DESCRIPTION
**Pull request by alpearce**
(Also check in generated metadata.)

We want to be explicit about making changes to rust-tuf that change the metadata it generates. go-tuf accomplishes this by containing checked-in keys and metadata generated by go-tuf and, in a unit test, asserts that it produces byte-for-byte equivalent metadata. That way, changes to the library that change generated metadata must include corresponding changes to the checked-in metadata to indicate the change in behavior. 

This patch includes the metadata generator and the first set of generated metadata. It uses the same keys as the go-tuf generator so that this metadata can also be used for comparison across implementations. 

Feel free to go wild on the review; this is the biggest hunk of Rust I've written and I'd appreciate a lot of comments.

### CODE REVIEW

File: tests/metadata/generate.rs [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+use tuf::crypto::{HashAlgorithm, KeyType, PrivateKey, SignatureScheme};
+use tuf::interchange::Json;
+use tuf::metadata::{
+    MetadataPath, MetadataVersion, Role, RootMetadataBuilder, SnapshotMetadataBuilder, TargetPath,
+    TargetsMetadataBuilder, TimestampMetadataBuilder, VirtualTargetPath,
+};
+use tuf::repository::{FileSystemRepository, FileSystemRepositoryBuilder, Repository};
+
+// TODO parametrize this better/find it within the build?
+const KEYS_PATH: &str = "./keys.json";
```
Comment (erickt) [inline tests/metadata/generate.rs:17]: This should be okay for now. I've been thinking about re-laying out the directory structure, so we could handle improving discovery of the keys.json file then.

File: tests/metadata/generate.rs [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    snapshot: Vec<Vec<TestKeyPair>>,
+    timestamp: Vec<Vec<TestKeyPair>>,
+}
+
+fn init_json_keys(path: &str) -> TestKeys {
+    let mut f = File::open(path).expect("failed to open keys file");
+    let mut contents = String::new();
+    f.read_to_string(&mut contents)
+        .expect("failed to read keys");
+    let keys: TestKeys = serde_json::from_str(&contents).expect("serde failed");
```
Comment (erickt) [inline tests/metadata/generate.rs]: You don't need to read this into a string. This should work:

```
let keys: TestKeys = serde_json::from_reader(f).expect("serde failed");
```

File: tests/metadata/generate.rs [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    keys.insert("timestamp", json_keys.timestamp[0][0].to_private_key());
+    keys
+}
+
+fn copy_repo(dir: &str, step: u8) {
+    let src = Path::new(dir)
+        .join((step - 1).to_string())
+        .join("repository");
+    let dst = Path::new(dir).join(step.to_string());
+    Command::new("cp")
```
Comment (erickt) [inline tests/metadata/generate.rs]: Could you use `/bin/cp`, just to make sure we don't pick up any old cp executable in the current directory? It should be okay if this only is runnable on a posix-y system, since this is a test script.

File: tests/metadata/generate.rs [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+fn init_role_keys(json_keys: &TestKeys) -> RoleKeys {
+    let mut keys = HashMap::new();
+    keys.insert("root", json_keys.root[0][0].to_private_key());
+    keys.insert("snapshot", json_keys.snapshot[0][0].to_private_key());
+    keys.insert("targets", json_keys.targets[0][0].to_private_key());
+    keys.insert("timestamp", json_keys.timestamp[0][0].to_private_key());
+    keys
+}
+
+fn copy_repo(dir: &str, step: u8) {
```
Comment (erickt) [inline tests/metadata/generate.rs:71]: Generally speaking I'd prefer this call some rust function to copy the directories, since we might not guarantee that `cp` is available. Eventually I'd like to turn this into a test library, so that we can add a test that will complain if we accidentally change the metadata format without regenerating the test metadata, like we're doing in [go-tuf](https://fuchsia.googlesource.com/third_party/go-tuf/+/refs/heads/master/client/interop_test.go#32).

Unfortunately, I'm guessing you discovered there's no standard library way for copying directories. We might be able to pull in a crate for this ([fs_extra](https://crates.io/crates/fs_extra) seems popular, but we need to check out it's quality), but for now I say we punt for now and solve this later for when we create this identity test.

So in short, could you add a `// TODO replace this with pure rust library so it's portable across OSes`?

File: tests/metadata/generate.rs [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+    repo.store_metadata(&root_path, &MetadataVersion::Number(version), &root)
+        .await
+        .unwrap();
+    repo.store_metadata(&root_path, &MetadataVersion::None, &root)
+        .await
+        .unwrap();
+}
+
+// adds a target and updates the non-root metadata files.
+async fn add_target(
```
Comment (erickt) [inline tests/metadata/generate.rs:121]: More evidence we need a [helper library for generating metadata](https://github.com/heartsucker/rust-tuf/issues/232).

File: tests/metadata/generate.rs [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+
+    // Timestamp doesn't require a version even in consistent_snapshot.
+    repo.store_metadata(&timestamp_path, &MetadataVersion::None, &timestamp)
+        .await
+        .unwrap();
+}
+
+async fn generate_repos(dir: &str, consistent_snapshot: bool) -> tuf::Result<()> {
+    // Create initial repo.
+    // TODO: @erickt, this impl is very tied to the json keys file. Do we need it to be more flexible?
```
Comment (erickt) [inline tests/metadata/generate.rs]: This should be fine for now.

File: tests/metadata/generate.rs [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+        // Initialize new repo and copy the files from the previous step.
+        let dir_i = Path::new(dir).join(i.to_string());
+        let repo = FileSystemRepositoryBuilder::new(dir_i)
+            .metadata_prefix(Path::new("repository"))
+            .targets_prefix(Path::new("repository").join("targets"))
+            .build()
+            .unwrap();
+        copy_repo(dir, i);
+
+        let root_signer: Option<PrivateKey> = match r {
```
Comment (erickt) [inline tests/metadata/generate.rs]: Nit: I think you can leave off `Option<PrivateKey>`.

**Review by erickt — CHANGES_REQUESTED** [review]
A few things other things:

* First, it looks like you hit a [test failure](https://github.com/heartsucker/rust-tuf/pull/237/checks?check_run_id=315130477). You can ignore the appveyor, that's [going away](https://github.com/heartsucker/rust-tuf/issues/235).
* Darn, I thought I added a mode FileSystemBuilder that lets us pretty print the metadata. That makes it much easier to review. I filed https://github.com/heartsucker/rust-tuf/issues/238 to track implementing this.

File: tests/metadata/generate.rs (hunk 1 above)
Comment (alpearce) [reply tests/metadata/generate.rs:17]: Ack.

File: tests/metadata/generate.rs (hunk 2 above)
Comment (alpearce) [reply tests/metadata/generate.rs]: Done.

File: tests/metadata/generate.rs (hunk 3 above)
Comment (alpearce) [reply tests/metadata/generate.rs]: Done.

File: tests/metadata/generate.rs (hunk 4 above)
Comment (alpearce) [reply tests/metadata/generate.rs:71]: Added TODO.

File: tests/metadata/generate.rs (hunk 6 above)
Comment (alpearce) [reply tests/metadata/generate.rs]: Ack.

File: tests/metadata/generate.rs (hunk 7 above)
Comment (alpearce) [reply tests/metadata/generate.rs]: Done.

**Comment by alpearce** [comment]
Updated:
  - @erickt's comments
  - use a constant expiration so metadata will be static
  - regenerate metadata with the above

**Comment by heartsucker** [comment]
Nice! We should fix #238 before merging this one otherwise reviews are going to be impossible.

**Comment by alpearce** [comment]
Borrowed the "linkify-metadata" script from go-tuf and regenerated symlinked metadata.

File: tests/metadata/generate.rs (hunk 5 above)
Comment (cavokz) [reply tests/metadata/generate.rs:121]: @alpearce is it ok for you if I attempt at the library part or is it something you would like to do?

**Comment by erickt** [comment]
A couple more things:

* I uploaded a pretty printing interchange in https://github.com/heartsucker/rust-tuf/pull/241, could you rebase your patch on top of this one and use it to regenerate the metadata? While you do that, can you squash all the patches together, then factor out the metadata into it's own patch? That should make it easier to review the metadata generation script from the actual metadata.
* I used your PR to test out that `JsonPretty` worked (thanks!) I was able to successfully parse step `0`, but I ran into a bug in step `1`. It is missing the `0` target. Instead, it should look like [this](https://github.com/heartsucker/rust-tuf/blob/develop/tests/interop/fuchsia-go-tuf-5527fe/consistent-snapshot-false/1/repository/targets.json).
* You don't need to do this now, but it'd be nice if we could migrate this script into a separate independent cargo toml. I'd be fine if you want us to do this in a future PR though to get this landed.

File: tests/metadata/generate.rs (hunk 5 above)
Comment (alpearce) [reply tests/metadata/generate.rs:121]: Fine with me!

**Comment by alpearce** [comment]
Prettified, squashed, separated metadata, fixed bug, added TODO in Cargo.toml.

**Review by erickt — APPROVED** [review]
